[BUG] Intermittent ValueError in test_struct_groupby_count

**Describe the bug**
Runs of test_struct_groupby_count intermittently throw exceptions like:
```
=================================== FAILURES ===================================
_ test_struct_groupby_count[Struct(not_null)(('a', Struct(('aa', Integer))))] __

self = Row(a=None), item = 'aa'

    def __getitem__(self, item):
        if isinstance(item, (int, slice)):
            return super(Row, self).__getitem__(item)
        try:
            # it will be slow when it has many fields,
            # but this will not be used in normal cases
>           idx = self.__fields__.index(item)
E           ValueError: 'aa' is not in list

```

**Steps/Code to reproduce bug**
Repeatedly running the test as  a dedicated Spark app per iteration
```bash
while (($? == 0)); do 
  ((iter++)) 
  NUM_LOCAL_EXECS=2 TEST_PARALLEL=1 LD_LIBRARY_PATH=$CONDA_PREFIX/lib \
  SPARK_HOME=~/dist/spark-3.1.1-bin-hadoop3.2 \
  ./integration_tests/run_pyspark_from_build.sh -k test_struct_groupby_count  
done
```

reproduces the error after a few iterations. 

Running multiple iterations in the same pytest app does not reproduce the error easily unless the first iteration already hits the bug

```bash 
NUM_LOCAL_EXECS=2 TEST_PARALLEL=1 LD_LIBRARY_PATH=$CONDA_PREFIX/lib \ 
SPARK_HOME=~/dist/spark-3.1.1-bin-hadoop3.2 \
./integration_tests/run_pyspark_from_build.sh -k test_struct_groupby_count --count=100 -x
```

**Expected behavior**
the test should work reliably

**Environment details (please complete the following information)**
local-cluster mode reproduces the failure

--- a/integration_tests/src/main/python/hash_aggregate_test.py
+++ b/integration_tests/src/main/python/hash_aggregate_test.py
@@ -605,22 +605,28 @@ def test_subquery_in_agg(adaptive, expr):
         conf = {"spark.sql.adaptive.enabled" : adaptive})
 
 
+def assert_single_level_struct(df):
+    first_level_dt = df.schema['a'].dataType
+    second_level_dt = first_level_dt['aa'].dataType
+    assert isinstance(first_level_dt, StructType)
+    assert isinstance(second_level_dt, IntegerType)
+
+
 @allow_non_gpu(any = True)
 @pytest.mark.parametrize('key_data_gen', [
     StructGen([
-        ('a', StructGen([
-            ('aa', IntegerGen(min_val=0, max_val=9))
-        ]))], nullable=False),
+        ('aa', IntegerGen(min_val=0, max_val=9)),
+    ], nullable=False),
     StructGen([
-        ('a', StructGen([
-            ('aa', IntegerGen(min_val=0, max_val=4)),
-            ('ab', IntegerGen(min_val=5, max_val=9)),
-        ]))], nullable=False),
+        ('aa', IntegerGen(min_val=0, max_val=4)),
+        ('ab', IntegerGen(min_val=5, max_val=9)),
+    ], nullable=False),
 ], ids=idfn)
 @ignore_order(local=True)
 def test_struct_groupby_count(key_data_gen):
     def group_by_count(spark):
         df = two_col_df(spark, key_data_gen, IntegerGen())
+        assert_single_level_struct(df)
         return df.groupBy(df.a).count()
     assert_gpu_and_cpu_are_equal_collect(group_by_count)
 
@@ -628,17 +634,18 @@ def test_struct_groupby_count(key_data_gen):
 @pytest.mark.parametrize('cast_struct_tostring', ['LEGACY', 'SPARK311+'])
 @pytest.mark.parametrize('key_data_gen', [
     StructGen([
-        ('a', IntegerGen(min_val=0, max_val=9)),
+        ('aa', IntegerGen(min_val=0, max_val=9)),
     ], nullable=False),
     StructGen([
-        ('a', IntegerGen(min_val=0, max_val=4)),
-        ('b', IntegerGen(min_val=5, max_val=9)),
+        ('aa', IntegerGen(min_val=0, max_val=4)),
+        ('ab', IntegerGen(min_val=5, max_val=9)),
     ], nullable=False)
 ], ids=idfn)
 @ignore_order(local=True)
 def test_struct_cast_groupby_count(cast_struct_tostring, key_data_gen):
     def _group_by_struct_or_cast(spark):
         df = two_col_df(spark, key_data_gen, IntegerGen())
+        assert_single_level_struct(df)
         return df.groupBy(df.a.cast(StringType())).count()
     assert_gpu_and_cpu_are_equal_collect(_group_by_struct_or_cast, {
         'spark.sql.legacy.castComplexTypesToString.enabled': cast_struct_tostring == 'LEGACY'
@@ -661,6 +668,7 @@ def test_struct_cast_groupby_count(cast_struct_tostring, key_data_gen):
 def test_struct_count_distinct(key_data_gen):
     def _count_distinct_by_struct(spark):
         df = gen_df(spark, key_data_gen)
+        assert_single_level_struct(df)
         return df.agg(f.countDistinct(df.a))
     assert_gpu_and_cpu_are_equal_collect(_count_distinct_by_struct)
 
@@ -681,6 +689,7 @@ def test_struct_count_distinct(key_data_gen):
 def test_struct_count_distinct_cast(cast_struct_tostring, key_data_gen):
     def _count_distinct_by_struct(spark):
         df = gen_df(spark, key_data_gen)
+        assert_single_level_struct(df)
         return df.agg(f.countDistinct(df.a.cast(StringType())))
     assert_gpu_and_cpu_are_equal_collect(_count_distinct_by_struct, {
         'spark.sql.legacy.castComplexTypesToString.enabled': cast_struct_tostring == 'LEGACY'

--- a/integration_tests/src/main/python/hash_aggregate_test.py
+++ b/integration_tests/src/main/python/hash_aggregate_test.py
@@ -15,12 +15,12 @@
 import pytest
 
 from asserts import assert_gpu_and_cpu_are_equal_collect, assert_gpu_and_cpu_are_equal_sql, assert_gpu_fallback_collect
-from conftest import get_non_gpu_allowed
 from data_gen import *
+from functools import reduce
 from pyspark.sql.types import *
 from marks import *
 import pyspark.sql.functions as f
-from spark_session import with_spark_session, is_before_spark_311
+from spark_session import is_before_spark_311
 
 _no_nans_float_conf = {'spark.rapids.sql.variableFloatAgg.enabled': 'true',
                        'spark.rapids.sql.hasNans': 'false',
@@ -611,6 +611,14 @@ def assert_single_level_struct(df):
     assert isinstance(first_level_dt, StructType)
     assert isinstance(second_level_dt, IntegerType)
 
+# Prevent value deduplication bug across rows and struct columns
+# https://github.com/apache/spark/pull/31778 by injecting
+# extra literal columns
+#
+def workaround_dedupe_by_value(df, num_cols):
+    col_id_rng = range(0, num_cols)
+    return reduce(lambda df, i: df.withColumn(f"fake_col_{i}", f.lit(i)), col_id_rng, df)
+
 
 @allow_non_gpu(any = True)
 @pytest.mark.parametrize('key_data_gen', [
@@ -627,7 +635,7 @@ def test_struct_groupby_count(key_data_gen):
     def group_by_count(spark):
         df = two_col_df(spark, key_data_gen, IntegerGen())
         assert_single_level_struct(df)
-        return df.groupBy(df.a).count()
+        return workaround_dedupe_by_value(df.groupBy(df.a).count(), 3)
     assert_gpu_and_cpu_are_equal_collect(group_by_count)
 
 

--- a/integration_tests/src/main/python/hash_aggregate_test.py
+++ b/integration_tests/src/main/python/hash_aggregate_test.py
@@ -605,6 +605,7 @@ def test_subquery_in_agg(adaptive, expr):
         conf = {"spark.sql.adaptive.enabled" : adaptive})
 
 
+# TODO support multi-level structs https://github.com/NVIDIA/spark-rapids/issues/2438
 def assert_single_level_struct(df):
     first_level_dt = df.schema['a'].dataType
     second_level_dt = first_level_dt['aa'].dataType

